bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/AbstractResourceIconProvider.java

```diff
@@ -0,0 +1,110 @@
+ * dimmablelight-82.png, dimmablelight-81.png, dimmablelight-80.png and return the first one it can find.
+ *
+ * @author Kai Kreuzer
+ *
+ */
+abstract public class AbstractResourceIconProvider implements IconProvider {
+
+    /**
+     * The OSGi bundle context
+     */
+    protected BundleContext context;
+
+    /**
+     * When activating the service, we need to keep the bundle context.
+     *
+     * @param context the bundle context provided through OSGi DS.
+     */
+    protected void activate(BundleContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public Set<IconSet> getIconSets() {
+        return getIconSets(null);
+    }
+
+    @Override
+    public Integer hasIcon(String category, String iconSetId, Format format) {
+        return hasResource(iconSetId, category.toLowerCase() + "." + format.toString().toLowerCase()) ? getPriority()
+                : null;
+    }
+
+    @Override
+    public InputStream getIcon(String category, String iconSetId, String state, Format format) {
+        String resourceWithState = category + ((state != null) ? "-" + state.toLowerCase() : "") + "."
+                + format.toString().toLowerCase();
+        if (hasResource(iconSetId, resourceWithState)) {
+            return getResource(iconSetId, resourceWithState);
+        } else {
+            // let's treat all percentage-based categories
+            try {
+                Double stateAsDouble = Double.valueOf(state);
+                if (stateAsDouble > 0 && stateAsDouble < 100) {
+                    for (int i = stateAsDouble.intValue(); i >= 0; i--) {
+                        String resourceWithNumberState = category + "-" + i + "." + format.toString().toLowerCase();
+                        if (hasResource(iconSetId, resourceWithNumberState)) {
+                            return getResource(iconSetId, resourceWithNumberState);
+                        }
+                    }
+                }
+            } catch (NumberFormatException e) {
+                // does not seem to be a number, so ignore it
+            }
+            return getResource(iconSetId, category + "." + format.toString().toLowerCase());
+        }
+    }
+
+    /**
+     * Provides the priority of this provider. A higher value will give this provider a precedence over others.
+     * 
+     * @return the priority as a positive integer
+     */
+    abstract protected Integer getPriority();
+
+    /**
+     * Provides the content of a resource for a certain icon set as a stream or null, if the resource does not exist.
+     *
+     * @param iconSetId the id of the icon set for which the resource is requested
+     * @param resourceName the name of the resource
+     * @return the content as a stream or null, if the resource does not exist
+     */
+    abstract protected InputStream getResource(String iconSetId, String resourceName);
```

Comment:
For another API (Sound resource API) I just searched for a good and abstract return value similar to InputStream, but which does not open the stream. Guava has a good class for it: ByteSource (http://docs.guava-libraries.googlecode.com/git-history/release/javadoc/com/google/common/io/ByteSource.html).

It has a lot of helper methods and serves as a good return type if you want to provide access to raw data without giving around opened streams. But unfortunately our Guava version is too old (we have 10.0.1 but we need 14.0).